Ability to pull in rules from remote repos

mkcomer opened the pull request:
**Please check if the PR fulfills these requirements**

- [x] The commit message follows our [guidelines](https://github.com/get-woke/woke/blob/main/CONTRIBUTING.md)
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Feature & Docs Update w/ feature


**What is the current behavior?** (You can also link to an open issue here)
This PR adds functionality to pull in remote rules from urls - https.

To test this PR locally:
go build
go run main.go -c "https://raw.githubusercontent.com/get-woke/woke/main/example.yaml" --debug

To test a 404 URL:
go run main.go -c "https://raw.githubusercontent.com/get-woke/woke/main/example" --debug



**What is the new behavior (if this is a feature change)?**



**Does this PR introduce a breaking change?** (What changes might users need to make due to this PR?)


**Other information**:

caitlinelfring commented on pkg/config/config_test.go:
```diff
@@ -155,6 +157,61 @@ func TestNewConfig(t *testing.T) {
+		assert.NoError(t, err)
+		assert.NotNil(t, c)
+		// delete downloaded file after successful test
+		os.Remove("downloadedRules.yaml")
```
I would suggest using `t.Cleanup` to handle cleanup like this https://pkg.go.dev/testing#T.Cleanup

mkcomer replied:
N/A after deleting download file functionality

caitlinelfring commented on pkg/config/config.go:
```diff
@@ -83,14 +88,73 @@ func (c *Config) ConfigureRules() {
+
+// downloads file from url to set filepath
+func DownloadFile(filepath string, url string) error {
+	var client = &http.Client{
```
The preferred way in go to define a variable would be `client := &http.Client{}`

mkcomer replied:
fixed in new code

caitlinelfring commented on pkg/config/config.go:
```diff
@@ -83,14 +88,73 @@ func (c *Config) ConfigureRules() {
+	var client = &http.Client{
+		Timeout: time.Second * 10,
+	}
+	ctx := context.Background()
+	req, _ := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
```
This is the same as `req, err := http.NewRequest(http.MethodGet, url, nil)` (it uses the background context). Also, please don't ignore the error here.

mkcomer replied:
fixed in new code

mkcomer commented:
I am merging in some updates which should resolve most of your above comments:

- Moved isValidURL to remote.go and updated unit tests for specific naming
- removed download functionality, rather a LoadRemoteConfig function that takes the body and unmarshalls the body to the config object 
- updated client := &http.Client{} and req, err := http.NewRequest(http.MethodGet, url, nil)


I would suggest an add on feature which allows you to set multiple remote configs, as I noticed that setting multiple local configs is not supported yet either.

Open to more feedback on this round of updates :)

mkcomer commented on pkg/config/config.go:
```diff
@@ -83,14 +88,73 @@ func (c *Config) ConfigureRules() {
+		return false
+	}
+
+	u, err := url.Parse(toTest)
```
ParseRequestURI() is our primary basic check, but it will allow strings like ‘Test: Test’ to pass (which we don’t want). We therefore combine it with a url parser to check that both a scheme (like http) and a host exist.

caitlinelfring replied:
Got it, thanks for clarifying!

mkcomer commented on pkg/config/config_test.go:
```diff
@@ -155,6 +157,61 @@ func TestNewConfig(t *testing.T) {
+}
+
+func Test_isValidURL(t *testing.T) {
+	t.Run("valid-url", func(t *testing.T) {
```
fixed this in updated code

mkcomer commented on pkg/config/config.go:
```diff
@@ -83,14 +88,73 @@ func (c *Config) ConfigureRules() {
+		// if fileName is a valid URL, we will download and set it to the config
+		log.Debug().Str("url", filename).Msg("Downloading file from")
+		// hardcoding this file and saving to root directory
+		downloadedFile := "downloadedRules.yaml"
```
fixed in new code

mkcomer commented on pkg/config/config.go:
```diff
@@ -83,14 +88,73 @@ func (c *Config) ConfigureRules() {
+	return true
+}
+
+// downloads file from url to set filepath
+func DownloadFile(filepath string, url string) error {
```
removed function in new code

mkcomer commented on pkg/config/config.go:
```diff
@@ -83,14 +88,73 @@ func (c *Config) ConfigureRules() {
 	return c, yaml.Unmarshal(yamlFile, &c)
 }
 
+// isValidUrl tests a string to determine if it is a valid URL or not
```
fixed in new code

caitlinelfring commented on pkg/config/config_test.go:
```diff
@@ -155,6 +157,18 @@ func TestNewConfig(t *testing.T) {
+		assert.NotNil(t, c)
+	})
+
+	t.Run("load-config-invalid-url", func(t *testing.T) {})
+	_, err := loadRemoteConfig("https://raw.githubusercontent.com/get-woke/woke/main/example")
+	assert.Error(t, err)
```
It doesn't look like you're running these in the testing function
```suggestion
	t.Run("load-config-invalid-url", func(t *testing.T) {
		_, err := loadRemoteConfig("https://raw.githubusercontent.com/get-woke/woke/main/example")
		assert.Error(t, err)
	})
```

caitlinelfring commented:
I know there was a GitHub actions outage earlier today, which is probably why the PR checks didnt run. They should run the next time you push but I'll be on the lookout. 

I'm hoping the codecov report is just out of date, but it looks like there's a dip in the coverage of `pkg/config/config.go`, I'd like to get that back up, if you can check if there are any uncovered lines once the PR check runs again.
